primary/src/block_waiter.rs

```diff
@@ -431,16 +435,47 @@ impl<PublicKey: VerifyingKey, SynchronizerHandler: Handler<PublicKey> + Send + S
         let mut futures = Vec::new();
         let mut ids = Vec::new();
 
+        // ensure payloads are synchronized for the found certificates
+        let found_certificates: Vec<Certificate<PublicKey>> = certificates
+            .clone()
+            .into_iter()
+            .filter(|(_, c)| c.is_some())
+            .map(|(_, c)| c.unwrap())
+            .collect();
+
+        let sync_result = self
+            .block_synchronizer_handler
+            .synchronize_block_payloads(found_certificates)
+            .await;
```

Comment:
Not super optimised strategy, as the whole process will block and wait until we have synchronized all the missing payloads.

Ideally, we could follow here a more stream based approach - so use directly the `block_synchronizer` which emits a message for each successful synchronized batch (best case when we do have then in near cache). However, doing this will require quite some refactoring and made the decision to not spend the time and prioritise delivery. Once we have some stress tests around this with worst case scenarios, we can refactor and improve.